File: tools/generate_assets.py
import math
import os
import struct
import zlib
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1] / "assets"


def write_png(path, width, height, rgb_func):
    raw = bytearray()
    for y in range(height):
        raw.append(0)
        for x in range(width):
            r, g, b = rgb_func(x, y, width, height)
            raw.extend(
                (
                    max(0, min(255, int(r))),
                    max(0, min(255, int(g))),
                    max(0, min(255, int(b))),
                )
            )

    def chunk(tag, data):
        return (
            struct.pack(">I", len(data))
            + tag
            + data
            + struct.pack(">I", zlib.crc32(tag + data) & 0xFFFFFFFF)
        )

    png = bytearray(b"\x89PNG\r\n\x1a\n")
    png += chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, 2, 0, 0, 0))
    png += chunk(b"IDAT", zlib.compress(bytes(raw), 9))
    png += chunk(b"IEND", b"")

    with open(path, "wb") as f:
        f.write(png)


def base_tex(x, y, width, height):
    gx = x / width
    gy = y / height
    mortar = 28 if (x // 32) % 2 == 0 else 18
    brick_line = y % 32 in (0, 1)
    vert_joint = (x + (16 if (y // 32) % 2 else 0)) % 64 in (0, 1)
    noise = (math.sin(x * 0.37) + math.cos(y * 0.29) + math.sin((x + y) * 0.11)) * 7
    base = 112 + int(20 * gx + 12 * gy + noise)
    r = base + 28
    g = base + 16
    b = base + 8
    if brick_line or vert_joint:
        r = g = b = mortar
    return r, g, b


def height_val(x, y):
    return (
        math.sin(x * 0.12) * 0.5
        + math.cos(y * 0.09) * 0.4
        + math.sin((x + y) * 0.04) * 0.35
    )


def normal_tex(x, y, width, height):
    sx1 = height_val(max(x - 1, 0), y)
    sx2 = height_val(min(x + 1, width - 1), y)
    sy1 = height_val(x, max(y - 1, 0))
    sy2 = height_val(x, min(y + 1, height - 1))
    dx = (sx2 - sx1) * 1.5
    dy = (sy2 - sy1) * 1.5
    nx, ny, nz = -dx, -dy, 1.0
    length = math.sqrt(nx * nx + ny * ny + nz * nz)
    nx, ny, nz = nx / length, ny / length, nz / length
    return (nx * 0.5 + 0.5) * 255, (ny * 0.5 + 0.5) * 255, (nz * 0.5 + 0.5) * 255


def gobo_tex(x, y, width, height):
    cx, cy = (width - 1) / 2, (height - 1) / 2
    dx, dy = x - cx, y - cy
    r = math.sqrt(dx * dx + dy * dy)
    a = math.atan2(dy, dx)
    spokes = math.sin(a * 8) > 0.2
    ring = 55 < r < 230
    blob = 1 if (spokes and ring) else 0
    soft = max(0.0, min(1.0, 1 - abs(r - 140) / 140))
    v = int((0.15 + 0.85 * (blob * soft)) * 255)
    return v, v, v


if __name__ == "__main__":
    os.makedirs(ROOT, exist_ok=True)
    write_png(str(ROOT / "sample_base_texture.png"), 1024, 1024, base_tex)
    write_png(str(ROOT / "sample_normal_map.png"), 1024, 1024, normal_tex)
    write_png(str(ROOT / "sample_gobo.png"), 512, 512, gobo_tex)
    print(f"Generated sample assets in {ROOT}")

File: tools/generate_assets.js
const fs = require("fs");
const path = require("path");
const zlib = require("zlib");

const ROOT = "C:\\Users\\ryryb\\Desktop\\LightingTexturePreviewer\\assets";

function crcTable() {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n += 1) {
    let c = n;
    for (let k = 0; k < 8; k += 1) c = (c & 1) ? (0xedb88320 ^ (c >>> 1)) : (c >>> 1);
    table[n] = c >>> 0;
  }
  return table;
}

const CRC_TABLE = crcTable();

function crc32(buf) {
  let c = 0xffffffff;
  for (let i = 0; i < buf.length; i += 1) {
    c = CRC_TABLE[(c ^ buf[i]) & 0xff] ^ (c >>> 8);
  }
  return (c ^ 0xffffffff) >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length, 0);
  const typeBuf = Buffer.from(type, "ascii");
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(Buffer.concat([typeBuf, data])), 0);
  return Buffer.concat([len, typeBuf, data, crc]);
}

function writePng(filePath, width, height, rgb) {
  const raw = Buffer.alloc((width * 3 + 1) * height);
  for (let y = 0; y < height; y += 1) {
    const rowStart = y * (width * 3 + 1);
    raw[rowStart] = 0;
    for (let x = 0; x < width; x += 1) {
      const [r, g, b] = rgb(x, y, width, height).map((v) => Math.max(0, Math.min(255, Math.round(v))));
      const i = rowStart + 1 + x * 3;
      raw[i] = r;
      raw[i + 1] = g;
      raw[i + 2] = b;
    }
  }
  const signature = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;
  ihdr[9] = 2;
  ihdr[10] = 0;
  ihdr[11] = 0;
  ihdr[12] = 0;
  const compressed = zlib.deflateSync(raw, { level: 9 });
  const out = Buffer.concat([
    signature,
    chunk("IHDR", ihdr),
    chunk("IDAT", compressed),
    chunk("IEND", Buffer.alloc(0))
  ]);
  fs.writeFileSync(filePath, out);
}

function heightVal(x, y) {
  return (
    Math.sin(x * 0.12) * 0.5 +
    Math.cos(y * 0.09) * 0.4 +
    Math.sin((x + y) * 0.04) * 0.35
  );
}

function baseTex(x, y, width, height) {
  const gx = x / width;
  const gy = y / height;
  const mortar = Math.floor(x / 32) % 2 === 0 ? 28 : 18;
  const brickLine = y % 32 === 0 || y % 32 === 1;
  const vertJoint = ((x + (Math.floor(y / 32) % 2 === 1 ? 16 : 0)) % 64 === 0) || ((x + (Math.floor(y / 32) % 2 === 1 ? 16 : 0)) % 64 === 1);
  const noise = (Math.sin(x * 0.37) + Math.cos(y * 0.29) + Math.sin((x + y) * 0.11)) * 7;
  const base = 112 + (20 * gx + 12 * gy + noise);
  if (brickLine || vertJoint) return [mortar, mortar, mortar];
  return [base + 28, base + 16, base + 8];
}

function normalTex(x, y, width, height) {
  const sx1 = heightVal(Math.max(x - 1, 0), y);
  const sx2 = heightVal(Math.min(x + 1, width - 1), y);
  const sy1 = heightVal(x, Math.max(y - 1, 0));
  const sy2 = heightVal(x, Math.min(y + 1, height - 1));
  const dx = (sx2 - sx1) * 1.5;
  const dy = (sy2 - sy1) * 1.5;
  let nx = -dx;
  let ny = -dy;
  let nz = 1.0;
  const len = Math.sqrt(nx * nx + ny * ny + nz * nz);
  nx /= len;
  ny /= len;
  nz /= len;
  return [(nx * 0.5 + 0.5) * 255, (ny * 0.5 + 0.5) * 255, (nz * 0.5 + 0.5) * 255];
}

function goboTex(x, y, width, height) {
  const cx = (width - 1) / 2;
  const cy = (height - 1) / 2;
  const dx = x - cx;
  const dy = y - cy;
  const r = Math.sqrt(dx * dx + dy * dy);
  const a = Math.atan2(dy, dx);
  const spokes = Math.sin(a * 8) > 0.2;
  const ring = r > 55 && r < 230;
  const blob = spokes && ring ? 1 : 0;
  const soft = Math.max(0, Math.min(1, 1 - Math.abs(r - 140) / 140));
  const v = (0.15 + 0.85 * (blob * soft)) * 255;
  return [v, v, v];
}

fs.mkdirSync(ROOT, { recursive: true });
writePng(path.join(ROOT, "sample_base_texture.png"), 1024, 1024, baseTex);
writePng(path.join(ROOT, "sample_normal_map.png"), 1024, 1024, normalTex);
writePng(path.join(ROOT, "sample_gobo.png"), 512, 512, goboTex);
console.log(`Generated assets in ${ROOT}`);
